Allow inserting a link with no text selected

noisysocks opened the pull request:
## Description
Closes #5960 and fixes #6900. When creating a link with no text selected, the link should be inserted with its text set to the URL.

## How has this been tested?
1. Place cursor within a text block and insert a link
2. Select some text and insert a link
3. Select some text and insert a link that opens in a new window
4. Select some text that contains an existing link and insert a link

## Screenshots <!-- if applicable -->
![links](https://user-images.githubusercontent.com/612155/40346336-ccff0426-5ddf-11e8-96c2-4a704f83eabf.gif)

noisysocks committed:
Default link text to the URL

When creating a link with no text selected, the link should be inserted
with its text set to the URL.

noisysocks commented on editor/components/rich-text/index.js:
```diff
@@ -789,12 +789,30 @@ export class RichText extends Component {
+						this.editor.execCommand( 'mceInsertLink', false, {
+							href,
+							...params,
+							'data-wp-temp-link': 1,
```
As far as I can tell, there's no way to get the node that `mceInsertLink` creates, so I borrowed this workaround from the classic editor.

https://github.com/WordPress/WordPress/blob/48077799d9d84b41d7ee1ea3a0054f0641813359/wp-includes/js/wplink.js#L396-L463

noisysocks commented on editor/components/rich-text/index.js:
```diff
@@ -789,12 +789,33 @@ export class RichText extends Component {
+						} );
+
+						// mceInsertLink doesn't return the anchor it creates, so query for it
+						// using a temporary attribute
```
I stole this idea from the classic editor.

https://github.com/WordPress/WordPress/blob/48077799d9d84b41d7ee1ea3a0054f0641813359/wp-includes/js/wplink.js#L418

iseulde commented on editor/components/rich-text/index.js:
```diff
@@ -789,12 +789,33 @@ export class RichText extends Component {
+					} );
+
+					// Place the cursor immediately after the newly inserted link
+					this.editor.selection.select( anchor );
```
Can't this be part of the transact callback?

iseulde commented on editor/components/rich-text/index.js:
```diff
@@ -789,12 +789,33 @@ export class RichText extends Component {
+						// mceInsertLink doesn't return the anchor it creates, so query for it
+						// using a temporary attribute
+						anchor = this.editor.getBody().querySelector( '[data-wp-temp-link]' );
+						anchor.removeAttribute( 'data-wp-temp-link' );
```
I'm not quite following why it is necessary to use the temporary attribute. Can't we insert the text and element at once? So I'm guessing when the selection is collapsed, we could use `editor.insertContent( '<a href="link">link</a>' )` or something like that?

noisysocks replied:
I wasn't having any luck with `isCollapsed()` yesterday but now when I try it it's fine—amazing what some sleep can do! I've reworked this in 4a02305ff1472d58a32e14fca6abbb01b0fea827 🙂

karmatosed reviewed (approved):
Looks good, thanks for iterating on this.

noisysocks committed:
Simplify link insertion

We can avoid the ugly data-wp-temp-link hack by using isCollapsed() and
insertContent().

noisysocks committed:
Fix links getting an unnecessary 'isadded' attribute

noisysocks committed:
Fix 'Open in new window' toggle

Fix regression where the 'Open in new window' does not work when adding
a new link.

noisysocks commented:
Thanks for reviewing! I reworked my approach in 7d35c924334c5e54e60cc8772734c41fa1ede16f and snuck in two bug fixes while I was at it:

1. 65c08e4de349bf3fc8a1ad2931cb00ddf017111a which fixes all of our `<a>` elements getting an unnecessary `isadding` attribute
2. fafe328a4a0540d45f061fd4df63ba265649d6db which fixes #6900

mtias commented:
Thanks for the improvement!
